csharp/src/Ice/SslTransceiver.cs

```diff
@@ -446,203 +445,117 @@ private void ReadCompleted(IAsyncResult result)
+            {
+                if (_engine.SecurityTraceLevel >= 1)
                 {
-                    //
-                    // The RemoteCertificateNotAvailable case does not appear to be possible
-                    // for an outgoing connection. Since .NET requires an authenticated
-                    // connection, the remote peer closes the socket if it does not have a
-                    // certificate to provide.
-                    //
-
-                    if (_incoming)
-                    {
-                        if (_verifyPeer > 1)
-                        {
-                            if (_engine.SecurityTraceLevel >= 1)
-                            {
-                                _communicator.Logger.Trace(_engine.SecurityTraceCategory,
-                                    "SSL certificate validation failed - client certificate not provided");
-                            }
-                            return false;
-                        }
-                        errors ^= (int)SslPolicyErrors.RemoteCertificateNotAvailable;
-                        message += "\nremote certificate not provided (ignored)";
-                    }
+                    _communicator.Logger.Trace(
+                        _engine.SecurityTraceCategory,
+                        "SSL certificate validation failed - Hostname mismatch");
                 }
+                return false;
+            }
+
+            X509Certificate2Collection? caCerts = _incoming ?
+                _engine.TlsServerOptions.ClientCertificateCertificateAuthorities :
+                _engine.TlsClientOptions.ServerCertificateCertificateAuthorities;
+
+            // If using custom certificate authorities and the peer provides a certificate,
+            // we rebuild the certificate chain with our custom chain policy.
+            if (caCerts != null && certificate != null)
+            {
+                chain = new X509Chain();
+                // We need to set this flag to be able to use a certificate authority from the extra store.
+                chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
 
-                if ((errors & (int)SslPolicyErrors.RemoteCertificateNameMismatch) > 0)
+                foreach (X509Certificate2 cert in caCerts)
                 {
-                    if (_engine.SecurityTraceLevel >= 1)
-                    {
-                        _communicator.Logger.Trace(_engine.SecurityTraceCategory,
-                            "SSL certificate validation failed - Hostname mismatch");
-                    }
-                    return false;
+                    chain.ChainPolicy.ExtraStore.Add(cert);
                 }
+                chain.Build(certificate as X509Certificate2);
+                errors |= SslPolicyErrors.RemoteCertificateChainErrors;
```

Comment:
Setting this error without checking the chain status is a bit unexpected. Why don't we check the chain status here to set this flag?

Reply:
Right we can check the chain first